File: codetandem/review.py
"""
AI-powered code review functionality.

This module handles generating prompts for AI code review and parsing
the AI's feedback.
"""

import json
from typing import Optional, Dict, Any
from dataclasses import dataclass

from codetandem.code_parser import CodeExtraction
from codetandem.providers.base import BaseAIProvider


@dataclass
class ReviewResult:
    """Represents the result of an AI code review."""

    success: bool
    feedback: str
    score: Optional[float] = None
    suggestions: Optional[list[str]] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary representation."""
        return {
            "success": self.success,
            "feedback": self.feedback,
            "score": self.score,
            "suggestions": self.suggestions or [],
        }


def build_review_prompt(
    code_extraction: CodeExtraction,
    module_context: Optional[str] = None,
    objective: Optional[str] = None,
) -> str:
    """
    Build a comprehensive prompt for AI code review.

    Args:
        code_extraction: Extracted code and TODO information
        module_context: Optional context about the current learning module
        objective: Optional specific learning objective being addressed

    Returns:
        Formatted prompt string for AI review
    """
    prompt_parts = []

    # Header
    prompt_parts.append("# Code Review Request")
    prompt_parts.append("")
    prompt_parts.append(
        "You are an expert code reviewer. Please review the following code submission."
    )
    prompt_parts.append("")

    # Context
    if module_context:
        prompt_parts.append(f"## Learning Module Context")
        prompt_parts.append(module_context)
        prompt_parts.append("")

    # Task description
    prompt_parts.append("## Task")
    if objective:
        prompt_parts.append(f"The user was asked to: {objective}")
    else:
        prompt_parts.append(f"TODO: {code_extraction.todo_text}")
    prompt_parts.append("")

    # Code submission
    prompt_parts.append("## User's Code Submission")
    prompt_parts.append(f"File: {code_extraction.file_path}")
    prompt_parts.append(
        f"Lines: {code_extraction.start_line}-{code_extraction.end_line}"
    )
    prompt_parts.append("")
    prompt_parts.append("```")
    prompt_parts.append(code_extraction.code)
    prompt_parts.append("```")
    prompt_parts.append("")

    # Review criteria
    prompt_parts.append("## Review Criteria")
    prompt_parts.append("")
    prompt_parts.append("Please evaluate the code based on:")
    prompt_parts.append("1. **Correctness**: Does the code fulfill the TODO/objective?")
    prompt_parts.append(
        "2. **Code Quality**: Is the code well-structured and readable?"
    )
    prompt_parts.append(
        "3. **Best Practices**: Does it follow language-specific conventions?"
    )
    prompt_parts.append(
        "4. **Learning Goals**: Does it demonstrate understanding of key concepts?"
    )
    prompt_parts.append("")

    # Response format
    prompt_parts.append("## Required Response Format")
    prompt_parts.append("")
    prompt_parts.append("Respond with a JSON object in this exact format:")
    prompt_parts.append("```json")
    prompt_parts.append("{")
    prompt_parts.append('  "success": true or false,')
    prompt_parts.append('  "feedback": "Detailed feedback message",')
    prompt_parts.append('  "score": 0-100 (numeric score),')
    prompt_parts.append('  "suggestions": ["suggestion 1", "suggestion 2", ...]')
    prompt_parts.append("}")
    prompt_parts.append("```")
    prompt_parts.append("")
    prompt_parts.append(
        "- Set `success` to `true` if the code correctly completes the task"
    )
    prompt_parts.append("- Set `success` to `false` if there are significant issues")
    prompt_parts.append("- Provide constructive `feedback` regardless of success")
    prompt_parts.append("- Include a `score` from 0-100 indicating code quality")
    prompt_parts.append("- List specific `suggestions` for improvement")
    prompt_parts.append("")
    prompt_parts.append(
        "**IMPORTANT**: Your entire response must be valid JSON. Do not include any text outside the JSON object."
    )

    return "\n".join(prompt_parts)


def parse_review_response(response_text: str) -> ReviewResult:
    """
    Parse the AI's review response.

    Args:
        response_text: Raw text response from AI

    Returns:
        ReviewResult object with parsed feedback
    """
    # Try to extract JSON from response
    response_text = response_text.strip()

    # Remove markdown code blocks if present
    if response_text.startswith("```json"):
        response_text = response_text[7:]
    elif response_text.startswith("```"):
        response_text = response_text[3:]

    if response_text.endswith("```"):
        response_text = response_text[:-3]

    response_text = response_text.strip()

    try:
        data = json.loads(response_text)

        return ReviewResult(
            success=data.get("success", False),
            feedback=data.get("feedback", "No feedback provided"),
            score=data.get("score"),
            suggestions=data.get("suggestions"),
        )
    except json.JSONDecodeError as e:
        # If JSON parsing fails, treat as a failure with the raw text as feedback
        return ReviewResult(
            success=False,
            feedback=f"Failed to parse AI response: {str(e)}\n\nRaw response:\n{response_text}",
            score=None,
            suggestions=None,
        )


async def review_code_with_ai(
    provider: BaseAIProvider,
    code_extraction: CodeExtraction,
    module_context: Optional[str] = None,
    objective: Optional[str] = None,
    temperature: float = 0.3,
) -> ReviewResult:
    """
    Submit code to AI for review.

    Args:
        provider: AI provider instance
        code_extraction: Extracted code and TODO information
        module_context: Optional context about the current learning module
        objective: Optional specific learning objective
        temperature: AI temperature setting (lower = more consistent)

    Returns:
        ReviewResult with AI's feedback
    """
    # Build the review prompt
    prompt = build_review_prompt(code_extraction, module_context, objective)

    try:
        # Get AI response
        response = await provider.chat(
            messages=[{"role": "user", "content": prompt}], temperature=temperature
        )

        # Parse the response
        result = parse_review_response(response.content)

        return result

    except Exception as e:
        return ReviewResult(
            success=False,
            feedback=f"Error during AI review: {str(e)}",
            score=None,
            suggestions=None,
        )


def format_review_feedback(review: ReviewResult) -> str:
    """
    Format review results for display to user.

    Args:
        review: ReviewResult object

    Returns:
        Formatted feedback string
    """
    lines = []

    # Status
    status = "✓ PASSED" if review.success else "✗ NEEDS IMPROVEMENT"
    lines.append(f"## Review Result: {status}")
    lines.append("")

    # Score
    if review.score is not None:
        lines.append(f"**Score**: {review.score}/100")
        lines.append("")

    # Feedback
    lines.append("### Feedback")
    lines.append(review.feedback)
    lines.append("")

    # Suggestions
    if review.suggestions:
        lines.append("### Suggestions for Improvement")
        for i, suggestion in enumerate(review.suggestions, 1):
            lines.append(f"{i}. {suggestion}")
        lines.append("")

    return "\n".join(lines)

File: src/utils/review.ts
/**
 * AI-powered code review functionality.
 *
 * Handles generating prompts for AI code review and parsing the AI's feedback.
 */

import { BaseAIProvider } from '../providers/base.js';
import type { CodeExtraction } from './code-parser.js';

export interface ReviewResult {
  success: boolean;
  feedback: string;
  score?: number;
  suggestions?: string[];
}

/**
 * Build a comprehensive prompt for AI code review.
 *
 * @param codeExtraction Extracted code and TODO information
 * @param moduleContext Optional context about the current learning module
 * @param objective Optional specific learning objective being addressed
 * @returns Formatted prompt string for AI review
 */
export function buildReviewPrompt(
  codeExtraction: CodeExtraction,
  moduleContext?: string,
  objective?: string
): string {
  const promptParts: string[] = [];

  // Header
  promptParts.push('# Code Review Request');
  promptParts.push('');
  promptParts.push('You are an expert code reviewer. Please review the following code submission.');
  promptParts.push('');

  // Context
  if (moduleContext) {
    promptParts.push('## Learning Module Context');
    promptParts.push(moduleContext);
    promptParts.push('');
  }

  // Task description
  promptParts.push('## Task');
  if (objective) {
    promptParts.push(`The user was asked to: ${objective}`);
  } else {
    promptParts.push(`TODO: ${codeExtraction.todoText}`);
  }
  promptParts.push('');

  // Code submission
  promptParts.push("## User's Code Submission");
  promptParts.push(`File: ${codeExtraction.filePath}`);
  promptParts.push(`Lines: ${codeExtraction.startLine}-${codeExtraction.endLine}`);
  promptParts.push('');
  promptParts.push('```');
  promptParts.push(codeExtraction.code);
  promptParts.push('```');
  promptParts.push('');

  // Review criteria
  promptParts.push('## Review Criteria');
  promptParts.push('');
  promptParts.push('Please evaluate the code based on:');
  promptParts.push('1. **Correctness**: Does the code fulfill the TODO/objective?');
  promptParts.push('2. **Code Quality**: Is the code well-structured and readable?');
  promptParts.push('3. **Best Practices**: Does it follow language-specific conventions?');
  promptParts.push('4. **Learning Goals**: Does it demonstrate understanding of key concepts?');
  promptParts.push('');

  // Response format
  promptParts.push('## Required Response Format');
  promptParts.push('');
  promptParts.push('Respond with a JSON object in this exact format:');
  promptParts.push('```json');
  promptParts.push('{');
  promptParts.push('  "success": true or false,');
  promptParts.push('  "feedback": "Detailed feedback message",');
  promptParts.push('  "score": 0-100 (numeric score),');
  promptParts.push('  "suggestions": ["suggestion 1", "suggestion 2", ...]');
  promptParts.push('}');
  promptParts.push('```');
  promptParts.push('');
  promptParts.push('- Set `success` to `true` if the code correctly completes the task');
  promptParts.push('- Set `success` to `false` if there are significant issues');
  promptParts.push('- Provide constructive `feedback` regardless of success');
  promptParts.push('- Include a `score` from 0-100 indicating code quality');
  promptParts.push('- List specific `suggestions` for improvement');
  promptParts.push('');
  promptParts.push(
    '**IMPORTANT**: Your entire response must be valid JSON. Do not include any text outside the JSON object.'
  );

  return promptParts.join('\n');
}

/**
 * Parse the AI's review response.
 *
 * @param responseText Raw text response from AI
 * @returns ReviewResult object with parsed feedback
 */
export function parseReviewResponse(responseText: string): ReviewResult {
  let text = responseText.trim();

  // Remove markdown code blocks if present
  if (text.startsWith('```json')) {
    text = text.substring(7);
  } else if (text.startsWith('```')) {
    text = text.substring(3);
  }

  if (text.endsWith('```')) {
    text = text.substring(0, text.length - 3);
  }

  text = text.trim();

  try {
    const data = JSON.parse(text) as Record<string, unknown>;

    return {
      success: data.success === true,
      feedback: typeof data.feedback === 'string' ? data.feedback : 'No feedback provided',
      score: typeof data.score === 'number' ? data.score : undefined,
      suggestions: Array.isArray(data.suggestions) ? (data.suggestions as string[]) : undefined,
    };
  } catch (e) {
    // If JSON parsing fails, treat as a failure with the raw text as feedback
    const error = e instanceof Error ? e.message : String(e);
    return {
      success: false,
      feedback: `Failed to parse AI response: ${error}\n\nRaw response:\n${responseText}`,
      score: undefined,
      suggestions: undefined,
    };
  }
}

/**
 * Submit code to AI for review.
 *
 * @param provider AI provider instance
 * @param codeExtraction Extracted code and TODO information
 * @param moduleContext Optional context about the current learning module
 * @param objective Optional specific learning objective
 * @param temperature AI temperature setting (lower = more consistent)
 * @returns ReviewResult with AI's feedback
 */
export async function reviewCodeWithAI(
  provider: BaseAIProvider,
  codeExtraction: CodeExtraction,
  moduleContext?: string,
  objective?: string,
  temperature: number = 0.3
): Promise<ReviewResult> {
  // Build the review prompt
  const prompt = buildReviewPrompt(codeExtraction, moduleContext, objective);

  try {
    // Get AI response
    const response = await provider.chat([{ role: 'user', content: prompt }], temperature);

    // Parse the response
    const result = parseReviewResponse(response.content);

    return result;
  } catch (e) {
    const error = e instanceof Error ? e.message : String(e);
    return {
      success: false,
      feedback: `Error during AI review: ${error}`,
      score: undefined,
      suggestions: undefined,
    };
  }
}

/**
 * Format review results for display to user.
 *
 * @param review ReviewResult object
 * @returns Formatted feedback string
 */
export function formatReviewFeedback(review: ReviewResult): string {
  const lines: string[] = [];

  // Status
  const status = review.success ? '✓ PASSED' : '✗ NEEDS IMPROVEMENT';
  lines.push(`## Review Result: ${status}`);
  lines.push('');

  // Score
  if (review.score !== undefined) {
    lines.push(`**Score**: ${review.score}/100`);
    lines.push('');
  }

  // Feedback
  lines.push('### Feedback');
  lines.push(review.feedback);
  lines.push('');

  // Suggestions
  if (review.suggestions && review.suggestions.length > 0) {
    lines.push('### Suggestions for Improvement');
    for (let i = 0; i < review.suggestions.length; i++) {
      lines.push(`${i + 1}. ${review.suggestions[i]}`);
    }
    lines.push('');
  }

  return lines.join('\n');
}
